ambry-account/src/main/java/com/github/ambry/account/HelixAccountService.java

```diff
@@ -0,0 +1,534 @@
+  static final String FULL_ACCOUNT_METADATA_PATH = "/account_metadata/full_data";
+  private static final String ZN_RECORD_ID = "full_account_metadata";
+  private final Logger logger = LoggerFactory.getLogger(getClass());
+  private final HelixPropertyStore<ZNRecord> helixStore;
+  private final Cache cache = new Cache();
+  private volatile boolean isOpen = false;
+
+  /**
+   * Constructor. It fetches the full account metadata set and caches locally.
+   * This call is blocking until it fetches all the {@link Account} metadata from {@link HelixPropertyStore}.
+   *
+   * @param storeConfig The config needed to start a {@link HelixPropertyStore}. Cannot be {@code null}.
+   * @param storeFactory The factory to generate a {@link HelixPropertyStore}. Cannot be {@code null}.
+   */
+  public HelixAccountService(HelixPropertyStoreConfig storeConfig, HelixPropertyStoreFactory<ZNRecord> storeFactory) {
+    if (storeConfig == null || storeFactory == null) {
+      throw new IllegalArgumentException(
+          "storeConfig=" + storeConfig + " and storeFactory=" + storeFactory + " cannot be null");
+    }
+    long startTimeMs = System.currentTimeMillis();
+    logger.info("Starting HelixAccountService with HelixPropertyStore host={} rootPath={}",
+        storeConfig.zkClientConnectString, storeConfig.rootPath);
+    helixStore = storeFactory.getHelixPropertyStore(storeConfig, null);
+    readFullAccountAndUpdateCache(FULL_ACCOUNT_METADATA_PATH);
+    isOpen = true;
+    logger.info("HelixAccountService started. {} accounts have been loaded, took time={}ms",
+        cache.getAccountTwoKeyMap().getAccounts().size(), System.currentTimeMillis() - startTimeMs);
+  }
+
+  @Override
+  public Account getAccountByName(String accountName) {
+    if (!isOpen) {
+      throw new IllegalStateException("AccountService is closed.");
+    }
+    if (accountName == null) {
+      throw new IllegalArgumentException("accountName cannot be null.");
+    }
+    return cache.getAccountByName(accountName);
+  }
+
+  @Override
+  public Account getAccountById(short id) {
+    if (!isOpen) {
+      throw new IllegalStateException("AccountService is closed.");
+    }
+    return cache.getAccountById(id);
+  }
+
+  @Override
+  public Collection<Account> getAllAccounts() {
+    if (!isOpen) {
+      throw new IllegalStateException("AccountService is closed.");
+    }
+    return cache.getAllAccounts();
+  }
+
+  /**
+   * {@inheritDoc}
+   * <p>
+   *   This call is blocking until it completes the operation of updating account metadata to {@link HelixPropertyStore}.
+   * </p>
+   * <p>
+   *   There is a slight chance that an {@link Account} could be updated successfully, even if its value was
+   *   set based on an outdated {@link Account}. This can be fixed after {@code generationId} is introduced
+   *   to {@link Account}.
+   * </p>
+   */
+  @Override
+  public boolean updateAccounts(Collection<Account> accounts) {
```

Comment:
Thanks for raising this question, and let me explain the complication here: So far there are two conflict checkings, and one version consistency checking. The first conflict checking is made against the local cache, once that is done, the second conflict checking will be made against the remote copy. If that also passes, the updates will be written to remote, where the version consistency checking happens. That is, assuming the updater thinks the remote ZNRecord's version is v1, but during update, the remote version is actually v2, the update operation will fail and retry. Retry includes fetching the remote copy, make a new update, and upload. This process will continue until it eventually succeeds.

In your scenario, when two threads are making update simultaneously, the two threads will succeed first in their local conflict checking and remote conflict checking. Then, assume thread one succeeds in writing the update, so the remote copy becomes `(1, a)`, thread 2 is doing update next. However, during the update operation, it sees version mismatch between the remote copy and its local copy, since is local copy is null for account id 1 before update, and the remote copy shows `(1, a)` before the update, therefore, it will fetch `(1, a)` to local and do update again. At this time, the second conflict check will fail, as there would be two accounts with the same name.